lib/puppet/type/file/content_uri.rb

```diff
@@ -0,0 +1,42 @@
+require 'net/http'
+require 'uri'
+
+module Puppet
+  Puppet::Type.type(:file).newparam(:content_uri) do
+
+    desc <<-'EOT'
+      A unique identifier for file content, which must be a URI which points
+      to a file in a module. This parameter allows the content of a file resource
+      to remain consistent regardless of code updates on the master.
+
+      The normal form of a `puppet:` URI is:
+
+      `puppet:///modules/<MODULE NAME>/files/<FILE PATH>`
+
+      Note that content_uri contains "files" in its path, unlike the source parameter.
+    EOT
+
+    validate do |path|
+      begin
+        uri = URI.parse(URI.escape(path))
+      rescue
+        self.fail Puppet::Error, "Could not understand URI #{path}: #{detail}", detail
+      end
+
+      self.fail "Cannot use opaque URLs '#{source}'" unless uri.hierarchical?
+      unless uri.scheme == "puppet"
+        self.fail "Must use URLs of type puppet as content URI"
+      end
+    end
+
+    def get_from_puppet_source(source_or_content, &block)
+      request = Puppet::Indirector::Request.new(:code_content, :find, resource[:content_uri], nil, :environment => resource.catalog.environment_instance)
```

Comment:
Do you mean: https://github.com/puppetlabs/puppet/tree/master/api?

Yeah, that makes *some* sense. It's a little tricky since the endpoint isn't going to be implemented in ruby at all, but it is going to be *used* from ruby.

@KevinCorcoran is the puppet-server repo gonna carry docs for this endpoint?